### ISSUE TITLE
Ragged edges

### ISSUE THREAD

**Issue by JoachimJohanssonCH** [casact/chainladder-python#167]
The various projection methods currently implemented don't seem to handle incomplete triangles out of the box. It would be very useful if they did. 

E.g. I'm looking at an example where I have some industry data for accident years 2008-2019, but I don't have any developments after 84 months, nor do I have developments for the first 24 months. I can still get an LDF, but to project, I have to manually iterate over each accident year and then apply my selected development pattern. 

It would be great if the projection methods worked without this manual workaround. Then I could e.g. also use stuff like BootstrapODPSample, which also doesn't work for years with an incomplete last diagonal. 

Related to this (it could be an addition) it would be great to have some imputation methods for missing values. These could e.g. be interpolations based on the observed development factors in each accident year.

**Comment by jbogaardt** [comment]
Hello @JoachimJohanssonCH ,

It would help me understand the request if wrapped in an example. I've worked an example up that I think follows your workflow, but not sure.  Does this approach eliminate the manual iteration on your end?

```python
import chainladder as cl

# Mock up some ragged "industry" data
industry_ragged = cl.load_sample('raa').loc[..., 36:84].dropna()

# Make factor selections from industry and convert to dict
industry_pattern_model = cl.Development(average='simple').fit(industry_ragged)
industry_patterns = dict(zip(industry_ragged.development[:-1], 
                             industry_pattern_model.ldf_.T.iloc[:, 0]))

# Make factor selections from customer data and convert to dict
# Only age 12 and 24 patterns will be used
some_customer = cl.load_sample('abc')
customer_pattern_model = cl.Development().fit(some_customer)
customer_pattern_model = dict(zip(some_customer.development[:-1], 
                                  customer_pattern_model.ldf_.T.iloc[:, 0]))

# Develop customer/industry blended pattern from dicts
blended_patterns = {age: industry_patterns.get(age, customer_pattern) 
                    for age, customer_pattern in customer_pattern_model.items()}

# Use blended patterns on customer with tail attaching at end of industry data
pipe = cl.Pipeline(
    steps=[
        ('dev', cl.DevelopmentConstant(patterns=blended_patterns, style='ldf')),
        ('tail', cl.TailCurve(attachment_age=84)),
        ('model', cl.Chainladder())],
)
pipe.fit(some_customer).named_steps.model.ibnr_
```

**Comment by JoachimJohanssonCH** [comment]
Hi,

I think this simple example illustrates what I was trying to do:
```
import pandas as pd 
import chainladder as cl 

# Create a triangle with one diagonal element missing
df = pd.DataFrame.from_dict({'AY':[2017, 2018, 2019, 2020, 2017, 2018, 2019, 2017, 2018],
                             'Year': [2018, 2019, 2020, 2021, 2019, 2020, 2021, 2020, 2021], 
                             'Incurred' : [1000, 1020, 1040, 1060, 1150, 1200, 1320, 1750, 1240]})

tri = cl.Triangle(df, origin='AY', development=['Year'], columns=['Incurred'])
dev = cl.Development().fit_transform(tri)
ult = cl.Chainladder().fit(dev)

# We don't get projections to ultimate for the first accident year. It would be great if we did:
ult.ultimate_

# If I wanted to use the BootstrapODPSample method, we can see that simulated values for the first AY are all zero:
boot = cl.BootstrapODPSample(n_sims=500).fit_transform(tri)
boot.iloc[0,0]
```

For the imputation idea, perhaps there's scope to do something similar to [sklearn](https://scikit-learn.org/stable/modules/impute.html) and use an LDF for imputing missing values?

**Comment by jbogaardt** [comment]
I think I understand. Since `Chainladder`, and by extension `BootstrapODPSample` are fundamentally pegged to the latest diagonal, it has to exist for all entries of that diagonal.  The `SimpleImputer` approach you describe seems like the solution.

I am cautious of modifying `Development`, `Chainladder` and `BootstrapODPSample` directly to support this since it seems like a big undertaking that will expand complexity of those estimators for the narrow "ragged edge" use case.  Whereas the `SimpleImputer` is self-contained, elegant, and for this specific purpose would solve all issues you mention.

Like sklearn, the imputer could have various imputation strategies.  Not sure what they should all be, but some of the more apparent ones are:
  1. Impute from preceding ldf
  2. Impute from the succeeding ldf
  3. Some blend of the two...?

Thoughts?

**Comment by JoachimJohanssonCH** [comment]
I think the three imputation options you mention all make sense. The “blend” could basically be the average of what you get when extrapolating forwards and backwards using a given LDF, e.g. to fill out a missing diagonal inside a triangle. 

It would be useful if the Imputer would work with BootstrapODPSample to first stochastically fill out the latest diagonal and then apply the BootstrapODPSample. Not sure if that would be feasible as part of a Pipeline? Maybe if BootstrapODPSample would accept an Imputation method as an optional parameter? That would then also be backwards compatible. 

I’ve not used the scikit-learn imputers that much, so I’m not sure if there are other ideas there which one may want to be able to leverage. Perhaps you know? Or if one could allow supplying a general function and rely on scipy.optimize.curve_fit when padding out the triangle. But such things could always come at a later stage - many issues could be addressed by a simple forward extrapolation using a supplied LDF.

**Comment by JoachimJohanssonCH** [comment]
I also often come across annual triangles, where the last diagonal is incomplete, e.g. representing only 6 or 9 mts development. If the imputation class could also re-state that diagonal, this would also be useful. 

Not entirely sure if there exists a standard approach for this.

**Comment by jbogaardt** [comment]
> I also often come across annual triangles, where the last diagonal is incomplete, e.g. representing only 6 or 9 mts development.

Working for a primary insurance carrier, I don't have experience with such messy data.  Any solution I can think up feels ugly or too heroric in its assumptions, and I'm not sure what to do here, i.e. I think I need help crafting this one.

**Comment by JoachimJohanssonCH** [comment]
Sure - I'll try to work out a proposal. Not sure I'll be able to get to it until after the 1/7 renewals, though.

**Comment by henrydingliu** [comment]
i'm closing this issue. all requested functionality is currently available within the package. 

[this](https://github.com/casact/chainladder-python/issues/167#issuecomment-858332644) just needs to be run twice, once normally, and another time on tri[valuation<=2020]. 

8-20-32-44 triangles are available within the package for [this](https://github.com/casact/chainladder-python/issues/167#issuecomment-861374075).
